### PR TITLE
[feat] legacy CRD compatibility

### PR DESCRIPTION
**Pull request by shaneutt**
**What this PR does / why we need it**:

This PR adds a workaround for an upstream problem in upgrading `apiextensions.k8s.io/v1beta1` versioned `CustomerResourceDefinitions` to `apiextensions.k8s.io/v1` versioned `CustomResourceDefinitions` in support of making it seamless to upgrade CRDs from KIC v1.x to the new CRDs in KIC v2.x.

For the full details, see the contributed `main.go` which includes documentation to justify its existence.

**Which issue this PR fixes**

Resolves https://github.com/Kong/kubernetes-ingress-controller/issues/1133
Supports https://github.com/Kong/kubernetes-ingress-controller/issues/1209

**PR Readiness Checklist**:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: railgun/hack/generators/manifests/main.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+		return fmt.Errorf("failed to patch crd in %s: %w; STDOUT=(%s) STDERR=(%s)",
+			path, err, strings.TrimSpace(stdout.String()), strings.TrimSpace(stderr.String()),
+		)
+	}
+
+	// controller-gen will add a YAML delimiter in the header, so we match that for consistency
+	patchedCRD := append([]byte("\n---\n"), stdout.Bytes()...)
+
+	// write the patched CRD back to the original file
+	return ioutil.WriteFile(path, patchedCRD, info.Mode().Perm())
```
Comment (mflendrich) [inline railgun/hack/generators/manifests/main.go:123]: YAGNI

```suggestion
	return nil
```

**Review by mflendrich — APPROVED** [review]

File: railgun/hack/generators/manifests/main.go (hunk 1 above)
Comment (shaneutt) [reply railgun/hack/generators/manifests/main.go:123]: So this is actually needed, the patch wont change the original file in place it only prints it to stdout. I did not find a flag to change this behavior.

File: railgun/config/crd/bases/configuration.konghq.com_kongclusterplugins.yaml [hunk 2]
```diff
@@ -14,6 +14,7 @@ spec:
     listKind: KongClusterPluginList
     plural: kongclusterplugins
     singular: kongclusterplugin
+  preserveUnknownFields: false
```
Comment (ccfishk) [inline railgun/config/crd/bases/configuration.konghq.com_kongclusterplugins.yaml:17]: Useful field. Do you have a test case covering this field ?

File: railgun/hack/generators/manifests/main.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+	// filter out any YAML files that don't contain a v1.CustomResourceDefinition
+	if !bytes.Contains(b, []byte(`kind: CustomResourceDefinition`)) || !bytes.Contains(b, []byte(`apiVersion: apiextensions.k8s.io/v1`)) {
+		return nil
+	}
+
+	// don't allow files that contain multiple objects (controller-gen wont be configured to emit these anyway)
+	yamlCRDs := bytes.Split(b, []byte(`\n---\n`))
+	if len(yamlCRDs) > 1 {
+		return fmt.Errorf("file %s contained multiple objects (%d) which this script doesn't yet support", path, len(yamlCRDs))
```
Comment (ccfishk) [inline railgun/hack/generators/manifests/main.go:103]: You want panic this err.

**Review by ccfishk — CHANGES_REQUESTED** [review]
left a few comment

File: railgun/config/crd/bases/configuration.konghq.com_kongclusterplugins.yaml (hunk 2 above)
Comment (shaneutt) [reply railgun/config/crd/bases/configuration.konghq.com_kongclusterplugins.yaml:17]: I didn't add a test case, because the field is statically false in v1 of `CustomResourceDefinition` so there's not really anything much to test, however in a later PR as per follow up and the scope of https://github.com/Kong/charts/issues/391 a chart test will be added to validate the upgrade.

File: railgun/hack/generators/manifests/main.go (hunk 1 above)
Comment (shaneutt) [reply railgun/hack/generators/manifests/main.go:123]: talked in zoom, good to resolve

File: railgun/hack/generators/manifests/main.go (hunk 3 above)
Comment (shaneutt) [reply railgun/hack/generators/manifests/main.go:103]: No I don't want this error to be a panic, this function takes an error, and any errors that occur will be caught by the file path walker which uses it to process each item.

File: railgun/config/crd/bases/configuration.konghq.com_kongclusterplugins.yaml (hunk 2 above)
Comment (ccfishk) [reply railgun/config/crd/bases/configuration.konghq.com_kongclusterplugins.yaml:17]: hmm, a follow up is good. Having been understand this PR will give user a smooth upgrade from version v1beta1 to v1, not aware we need other supplement.

File: railgun/hack/generators/manifests/main.go (hunk 3 above)
Comment (ccfishk) [reply railgun/hack/generators/manifests/main.go:103]: Do you mean even the YAML is empty, the CRD migration could still proceed ? I dont believe that is what you want.

File: CHANGELOG.md [hunk 4]
```diff
@@ -55,6 +55,13 @@
 
 #### Under the hood
 
+- New `v1` versions of `CustomResourceDefinitions` introduced for KIC 2.0 are now
+  backwards compatible with the previous `v1beta1` CRD definitions. In practice
```
Comment (ccfishk) [inline CHANGELOG.md:59]: "backwards compatability" do you men "migrate from v1beta1 to v1" ? Or support both version ?
Kindly add some clarity.

File: CHANGELOG.md [hunk 5]
```diff
@@ -55,6 +55,13 @@
...
+  the upgrade process should be seamless for end-users (e.g. `kubectl apply -f`).
```
Comment (ccfishk) [inline CHANGELOG.md:60]: kubectl apply -f what ? what resources are contained within the yaml ?

File: railgun/hack/generators/manifests/main.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+// -----------------------------------------------------------------------------
+// Vars
+// -----------------------------------------------------------------------------
+
+var (
+	// directory that will be used for walking the filesystem to find CRDs.
+	directory string
+
+	// patchfile is the patch that will be used for each CRD file
+	patchfile = "config/crd/patches/upgrade_compat.yaml"
```
Comment (ccfishk) [inline railgun/hack/generators/manifests/main.go:23]: Is this patch or upgrade ? the function in this PR more like crd/upgrade

File: railgun/hack/generators/manifests/main.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+// -----------------------------------------------------------------------------
+
+func main() {
+	// handle arguments
+	flag.StringVar(&directory, "directory", ".", "which directory to find CRDs in")
+	flag.Parse()
+
+	// find all the YAML CRDs in the provided directory and patch them
+	if err := filepath.Walk(directory, processFile); err != nil {
+		fmt.Fprintf(os.Stderr, "Error processing files in %s: %s\n", directory, err)
```
Comment (ccfishk) [inline railgun/hack/generators/manifests/main.go:70]:
```suggestion
		fmt.Fprintf(os.Stderr, "Error processing files in %s: %w\n", directory, err)
```

File: railgun/hack/generators/manifests/main.go [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+	}
+
+	// don't allow files that contain multiple objects (controller-gen wont be configured to emit these anyway)
+	yamlCRDs := bytes.Split(b, []byte(`\n---\n`))
+	if len(yamlCRDs) > 1 {
+		return fmt.Errorf("file %s contained multiple objects (%d) which this script doesn't yet support", path, len(yamlCRDs))
+	}
+
+	// generate a patch command to ensure "preserveUnknownFields" is set to false
+	stdout, stderr := new(bytes.Buffer), new(bytes.Buffer)
```
Comment (ccfishk) [inline railgun/hack/generators/manifests/main.go:107]: dont check stderr ?

File: railgun/hack/generators/manifests/main.go [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+	patchCMD := exec.Command("kubectl", "patch", "--type", "merge", "--local", "-o", "yaml", "-f", path, "--patch-file", patchfile)
+	patchCMD.Stdout = stdout
+	patchCMD.Stderr = stderr
+
+	// run the patcher and make sure there are no errors
+	if err := patchCMD.Run(); err != nil {
+		return fmt.Errorf("failed to patch crd in %s: %w; STDOUT=(%s) STDERR=(%s)",
```
Comment (ccfishk) [inline railgun/hack/generators/manifests/main.go:114]: migration, upgrade makes more sense ?

**Review by ccfishk — CHANGES_REQUESTED** [review]
comments about error, wording

File: railgun/config/crd/bases/configuration.konghq.com_kongclusterplugins.yaml (hunk 2 above)
Comment (shaneutt) [reply railgun/config/crd/bases/configuration.konghq.com_kongclusterplugins.yaml:17]: Sounds good, let's follow up :+1:

File: railgun/hack/generators/manifests/main.go (hunk 3 above)
Comment (shaneutt) [reply railgun/hack/generators/manifests/main.go:103]: > even the YAML is empty

The checks above which check for specific contents check whether or not the file is empty (see lines 80-100).

This line only checks whether the file has _multiple objects defined_ (as noted by the comment above and the error output) and simply fails (doesn't support) those kinds of files.

However there is no need to panic, as this error will be caught, processed, and a non-zero exit code will be thrown.

File: CHANGELOG.md (hunk 4 above)
Comment (shaneutt) [reply CHANGELOG.md:59]: backwards compatibility, meaning v1 is now compatible backwards [with v1beta1]

File: CHANGELOG.md (hunk 5 above)
Comment (shaneutt) [reply CHANGELOG.md:60]: You're right, let's clarify:
```suggestion
  the upgrade process should be seamless for end-users (e.g. `kubectl apply -f <NEW CRDS>`).
```

File: railgun/hack/generators/manifests/main.go (hunk 6 above)
Comment (shaneutt) [reply railgun/hack/generators/manifests/main.go:23]: this is a patch which enables CRD upgrade compatibility. "patch" in this case is a specific type of YAML file which this file conforms to.

File: railgun/hack/generators/manifests/main.go (hunk 7 above)
Comment (shaneutt) [reply railgun/hack/generators/manifests/main.go:70]: In this case the error wrapping is not necessary because this error has no caller in our codebase: it can't be captured it's being printed straight to STDERR before we exit the program.

File: railgun/hack/generators/manifests/main.go (hunk 8 above)
Comment (shaneutt) [reply railgun/hack/generators/manifests/main.go:107]: `stderr` and `err` are basically mutually exclusive. If an error occurs we catch it below in `err` and then we print out any captured stderr output in the error.

File: railgun/hack/generators/manifests/main.go (hunk 9 above)
Comment (shaneutt) [reply railgun/hack/generators/manifests/main.go:114]: In this context the topical verb is "patch" because all this file does is use `kubectl patch` at the end of the day, so I don't think `migration` or `upgrade` makes any sense because those aren't things we're actually doing as a part of this script.

File: CHANGELOG.md (hunk 4 above)
Comment (ccfishk) [reply CHANGELOG.md:59]: how do user expect from the "backwards compatibility" ? do you mind provide more clear/concrete statement, both v1 & v1beta1, from v1beta1 -> v1, from v1-beta1 ?

File: railgun/hack/generators/manifests/main.go (hunk 6 above)
Comment (ccfishk) [reply railgun/hack/generators/manifests/main.go:23]: Generically, software patch means operation fix, update, improve. Aren't you  doing both version compatible.

File: CHANGELOG.md (hunk 4 above)
Comment (shaneutt) [reply CHANGELOG.md:59]: Understood, let me add a small diagram to that effect, let me know if you feel this helps:
```suggestion
  backwards compatible with the previous `v1beta1` CRD definitions (i.e. `v1beta1 -> v1`
  upgrades of KIC's CustomResourceDefinitions now work fully automatically). In practice
```

File: railgun/hack/generators/manifests/main.go (hunk 7 above)
Comment (ccfishk) [reply railgun/hack/generators/manifests/main.go:70]: Well, print an object error into string would lose the log meaning in this case. maybe you want err.String ?

File: railgun/hack/generators/manifests/main.go (hunk 6 above)
Comment (shaneutt) [reply railgun/hack/generators/manifests/main.go:23]: I'm sorry I may need further help to understand the problem, could you please add a suggestion of the name you would prefer?

File: railgun/hack/generators/manifests/main.go (hunk 8 above)
Comment (ccfishk) [reply railgun/hack/generators/manifests/main.go:107]: Line 107 presents stderr, it'd better check error in this case, and this error is thrown by new function.
Following code and error might be related, but definitely different function/api/operation call.

File: railgun/hack/generators/manifests/main.go (hunk 7 above)
Comment (shaneutt) [reply railgun/hack/generators/manifests/main.go:70]: With an `error` type object there's no need to explicitly call `String` when using the string formatter `%s`, this will not mangle the error contents.

Example: https://play.golang.org/p/tLUs1QoGdhA

File: railgun/hack/generators/manifests/main.go (hunk 8 above)
Comment (shaneutt) [reply railgun/hack/generators/manifests/main.go:107]: We don't need to check the contents of `stderr` because we make a binary decision when failures occur: we fail the program and print the reason for the failure if one occurs. However, we do print the stderr when we fail on line 114 so that the end-user/caller can decide how to proceed.

File: railgun/hack/generators/manifests/main.go (hunk 6 above)
Comment (ccfishk) [reply railgun/hack/generators/manifests/main.go:23]: upgrade
migrate

File: railgun/hack/generators/manifests/main.go (hunk 9 above)
Comment (ccfishk) [reply railgun/hack/generators/manifests/main.go:114]: sorry, but which file/line is executing "kubectl patch " ?

File: railgun/hack/generators/manifests/main.go (hunk 7 above)
Comment (ccfishk) [reply railgun/hack/generators/manifests/main.go:70]: convert a struct to string makes the print out not like  a meaningful string. thats the awkard part. 
It is not a blocker to merge.

File: railgun/hack/generators/manifests/main.go (hunk 6 above)
Comment (shaneutt) [reply railgun/hack/generators/manifests/main.go:23]: I'm sorry, could you give a very prescriptive suggestion, such as suggesting the full path to the file as a github suggestion? I'm afraid it's still not clear to me _where_ to change this language.

File: railgun/hack/generators/manifests/main.go (hunk 7 above)
Comment (shaneutt) [reply railgun/hack/generators/manifests/main.go:70]: `error` is not a struct, it is an interface type, and the examples above show that this will result in a meaningful string. Since you've pointed out this is a non-blocker I will resolve it, let me know if there's something I've missed though.

File: railgun/hack/generators/manifests/main.go (hunk 9 above)
Comment (shaneutt) [reply railgun/hack/generators/manifests/main.go:114]: line `113` is executing `kubectl patch` (line `108` defines it)

File: railgun/hack/generators/manifests/main.go (hunk 6 above)
Comment (ccfishk) [reply railgun/hack/generators/manifests/main.go:23]: Q: "could you please add a suggestion of the name you"
A: upgrade, migrate

File: railgun/hack/generators/manifests/main.go (hunk 8 above)
Comment (ccfishk) [reply railgun/hack/generators/manifests/main.go:107]: It is very unusual the code does not check stderr.

File: railgun/hack/generators/manifests/main.go (hunk 9 above)
Comment (ccfishk) [reply railgun/hack/generators/manifests/main.go:114]: Line 108 does patch 
Its very expensive executing "kubectl patch" within a golang module and it is not a test case, could you please change to use client-go library  instead ?

**Review by ccfishk — CHANGES_REQUESTED** [review]
left comments

File: railgun/hack/generators/manifests/main.go (hunk 6 above)
Comment (mflendrich) [reply railgun/hack/generators/manifests/main.go:23]: Drive-by comment to add some more context and a specific proposed action stemming from it:

@ccfishk, "patch" here refers to the act of "patching" a YAML file with a "YAML patch". See [jsonpatch](http://jsonpatch.com/) as an example use of the term "patch" for overwriting a part of a JSON object.

This PR implements a program that takes the v1 CRDs generated by kubebuilder and **patches** them with a change that makes them suitable as an upgrade/migration from v1beta1 CRDs.

In other words, kubebuilder generates v1 CRDs that don't apply cleanly over v1beta1 CRDs.

The v1 CRDs patched with this YAML patch, however, can be applied over v1beta1 CRDs.

Does that make sense now? In other words, I think that the word "patch" here is appropriate and we should keep it as-is. Because "patching" does not refer to "patched software" but to "a YAML file with a patch applied to it".

File: railgun/hack/generators/manifests/main.go (hunk 8 above)
Comment (mflendrich) [reply railgun/hack/generators/manifests/main.go:107]: @ccfishk could you please propose a code suggestion here? I think I don't understand the concern.

File: railgun/hack/generators/manifests/main.go (hunk 9 above)
Comment (mflendrich) [reply railgun/hack/generators/manifests/main.go:114]: @ccfishk, see https://github.com/Kong/kubernetes-ingress-controller/pull/1559#discussion_r671287102 which is the context behind the decision to use `kubectl patch`.

Note that we perform `kubectl patch` locally, without interaction with any kubernetes API server. Our intention is to only edit a YAML file on the disk by adding a field to it. The PR linked above is the reason why we opted against deserializing and serializing the CRD YAML and editing it programmatically in Go.

My suggestion is to keep `kubectl patch` here as-is, so that we don't need to maintain the patching logic.

**Review by mflendrich — APPROVED** [review]
To help clarify misunderstanding about the role of YAML patching in this PR (and in the code generator being implemented here), added some comments.

To me this is ready to go as is, but let's get on the same page before merging.

File: railgun/hack/generators/manifests/main.go (hunk 8 above)
Comment (ccfishk) [reply railgun/hack/generators/manifests/main.go:107]: if strings.contains(stderr, some-error-keyword(i.e"err")){
  return fmt.Errorf("failed to ")
}
does it make sense to you ?

File: railgun/hack/generators/manifests/main.go (hunk 9 above)
Comment (ccfishk) [reply railgun/hack/generators/manifests/main.go:114]: @shaneutt  and Michal, lets leave as it is.
Lets get this merged asap.

**Review by ccfishk — APPROVED** [review]
